pi5/hw_pwm.go

```diff
@@ -166,6 +195,99 @@ func (pwm *pwmDevice) callGPIOReadall() {
+
+		However, you only need 1 byte to represent all the modes because
+			1 byte = 8 bits
+			8 modes can be represented by 3 bits
+
+		In practice, only 1 of the mode bytes is changed and the rest are just 0xff's.
+		The index of that byte in the entire chunk of pin data is pinBytes[4].
+	*/
+
+	altModeIndex := 4
+
+	pinAddress, err := getGPIOPinAddress(GPIONumber)
+	if err != nil {
+		return fmt.Errorf("error getting gpio pin address: %w", err)
+	}
+
+	// find pin data within virtual page; retrieve the 4th byte from the pin data
+	vPage := *(pwm.gpioChipVPage)
+	pinBytes := vPage[pinAddress : pinAddress+pinDataSize]
+	altModeByte := pinBytes[altModeIndex]
+
+	// We keep the left half (4 bits) of the byte preserved, only modifying the right half
+	// Preserve Left Side of Byte using this Mask
+	leftSideMask := byte(0xf0)
+
+	// Set Right Side of Byte ONLY using this Mask. Ensures left side cannot be overwritten
+	rightSideMask := byte(0x0f)
+
+	// Set new mode via write to correct while protecting previous other settings
+	newAltModeByte := (altModeByte & leftSideMask) | (newMode & rightSideMask)
+	pinBytes[altModeIndex] = newAltModeByte
+
+	return nil
+}
+
+/*
+For all pins belonging to the same bank, pin data is stored contiguously and in 8 byte chunks.
+For a given pin, this method determines:
+ 1. which bank the pin belongs to
+ 2. the starting address of its 8 byte data chunk
+*/
+func getGPIOPinAddress(GPIONumber int) (int64, error) {
+
+	// Regarding the header: I (Maria) am unsure about what is stored here. It might just be GPIO 0.
+	// In that case, banks1 & 2 would be wrong for including the header in offset calcs.
+	bankHeaderSizeBytes := 8 // 8 bytes of either header data assosciated with a bank
+
+	if !(1 <= GPIONumber && GPIONumber <= maxGPIOPins) {
+		return -1, errors.New("pin is out of bank range")
+	}
+
+	pinBankOffset := (GPIONumber - 1) // 1 is the number of starting GPIO Pin in Bank0 (GPIO #1)!
+	pinAddressOffset := bank0Offset + bankHeaderSizeBytes + (pinBankOffset * pinDataSize)
+	return int64(pinAddressOffset), nil
+}
+
+/*
+This function uses pwm.line to determine what GPIO Pin it needs to set to the inputted mode
+Each pwm line corresponds to a GPIO Pin. The pi5 mapping is:
+
+	PWM Line 0 -> GPIO 12
+	PWM Line 1 -> GPIO 13
+	PWM Line 2 -> GPIO 18
+	PWM Line 3 -> GPIO 19
+
+Other mappings for different pis can be found here:  https://pypi.org/project/rpi-hardware-pwm/#modal-close
+
+Use the writeToPinModeByte to set the mode to PWM or GPIO using this helper method. Pin Mode will either be 'PWMMode' or 'GPIO'
+
+TODO: Make writing to GPIO Pins generalizeable, and not dependent on the pwm following this exact structure
+*/
+func (pwm *pwmDevice) SetPinMode(pinMode byte) (err error) {
+	switch pwm.line {
+	case 0:
+		err = pwm.writeToPinModeByte(12, pinMode)
+	case 1:
+		err = pwm.writeToPinModeByte(13, pinMode)
+	case 2:
+		err = pwm.writeToPinModeByte(18, pinMode)
+	case 3:
+		err = pwm.writeToPinModeByte(19, pinMode)
+	}
```

Comment:
I'd put in a default case that has errors about an unknown PWM line. Otherwise, someone in the future will reuse this function on a different board and we'll silently stop doing stuff while claiming everything is working normally.

Reply:
i had this before but based on previous comments I thought you said I should remove it? I'll add it back in!